src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -742,5 +748,100 @@ private void UpdateCacheExpirationTime(KeyValueWatcher changeWatcher, bool succe
 
             changeWatcher.CacheExpires = DateTimeOffset.UtcNow.Add(cacheExpirationTime);
         }
+
+        private async Task ExecuteWithFailOverPolicyAsync(Func<ConfigurationClient, Task> funcToExecute, CancellationToken cancellationToken = default)
+        {
+            using IEnumerator<ConfigurationClient> clientEnumerator = _configClientProvider.GetClients().GetEnumerator();
+
+            // Guaranteed to return at least 1 element
+            Debug.Assert(clientEnumerator.MoveNext());
+
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                try
+                {
+                    await funcToExecute(clientEnumerator.Current).ConfigureAwait(false);
+
+                    _configClientProvider.UpdateClientStatus(clientEnumerator.Current, successful: true);
+
+                    break;
+                }
+                catch (AggregateException ae)
+                {
+                    if (!ShouldFailoverForException(ae))
+                    {
+                        throw;
+                    }
+
+                    _configClientProvider.UpdateClientStatus(clientEnumerator.Current, successful: false);
+
+                    if (!clientEnumerator.MoveNext())
+                    {
+                        throw;
+                    }
+                }
+                catch (RequestFailedException e)

```

Comment:
Looks like the handling code for RequestFailedException is exactly same as the code for AggregateException. Can we avoid duplicated code?

Reply:
two options!
1. We catch `Exception` and put `when` clause to restrict only catching `AggregateException` and `RequestFailedException`, but Jimmy mentioned we shouldn't catch the `Exception` as it is very broad.
2. We could create a private method to handle the exception in two separate catch blocks, but the `throw` inside the catch block would go into the private method and may pollute the stack trace.
Not sure which one would be better one.

Reply:
You could have the catch look like

```
catch (ae)
{
    if (!TryFailover(clientEnumerator, ae)
    {
      throw;
    }
}
```

The private method doesn't need to throw, just return false. STill some code duplication, but less.

Reply:
Done!